### DISCUSSION TITLE
Wiznet W5500 on the ESP32 (2022 edition)

### DISCUSSION THREAD

**Discussion by victorallume** [micropython/micropython#9471] [Libraries & Drivers]
I'm trying to get the Wiznet W5x00 running on the ESP32. I note there are a few discussion in the forums on this from previous years, but things seem to be a bit different now. I'm hoping to get this working using the C-based drivers, not the pure-python implementation mentioned in [this thread](https://forum.micropython.org/viewtopic.php?f=2&t=5851&p=36361)

So I've tried adding this to `boards/GENERIC_S3/mpconfigboard.cmake`:
```
set(MICROPY_PY_LWIP ON)
set(MICROPY_PY_NETWORK_WIZNET5K W5500)
```
but compilation fails due to the LWIP option there.

So I'm now trying to understand where LWIP sits in the ESP32 port. I understand it's part of the ESP-IDF - does that mean it's not exposed to micropython in the same way as in other ports? What would be my options here - would I need to try to get wiznet support into the ESP-IDF, or is there a way to use the existing wiznet drivers in `extmod`? It seems the wiznet library can work without LWIP, but if I compile it without this, I can't seem to access it through the `network` module. I would prefer to have wiznet and wifi accessible through the standard `network` module (though not at the same time). 

While using the LAN8720 PHY would be a 'cleaner' option, there are supply-chain issues with these at present (no industrial-rate stock till late 2023) so I'm looking at other options; the wiznet seems very available.

**Comment by dpgeorge** [comment]
> So I'm now trying to understand where LWIP sits in the ESP32 port. I understand it's part of the ESP-IDF - does that mean it's not exposed to micropython in the same way as in other ports?

That is correct.  lwIP is integrated into the IDF and we only really have access to the top layer of it, the functions to open and use sockets.

> What would be my options here - would I need to try to get wiznet support into the ESP-IDF, or is there a way to use the existing wiznet drivers in extmod

To integrate Wiznet into lwIP would require either (1) getting official support for Wiznet in the IDF, likely that would take a long time; (2) finding out a way to hook into the bottom layers of lwIP on the ESP32, likely that will be difficult; (3) add a second copy of lwIP to the ESP32 build specifically for Wiznet, likely that will cause lots of issues coexisting with the built-in lwIP.

The easiest way to use Wiznet would be without lwIP, using the TCP/IP stack on the Wiznet.

> While using the LAN8720 PHY would be a 'cleaner' option, there are supply-chain issues with these at present

The ESP32 supports quite a few different LAN PHYs (eg IP101, RTL8201, KSZ8041) and it would be *much* simpler to just use one of these if you can source them.

**Reply by victorallume** [reply]
Thanks. If I want to use wiznet without LWIP, what do I need to do (either in compiling micropython, or using it?) I've tried setting `set(MICROPY_PY_NETWORK_WIZNET5K W5500)` in `mpconfigboard.cmake` but `WIZNET` doesn't appear in the `network` module at runtime - should I be looking elsewhere, or do I need to change something else in the build?

**Reply by jimmo** [reply]
We don't have any support for this on ESP32. You've got to use the ESP IDF stack (for now).

**Reply by jimmo** [reply]
In more detail -- the ESP32 port uses its own implementation of `socket` and `network`, which only know how to talk to the IDF functionality. To make this work, you'd have to make ESP32 used the shared implementation of modnetwork (i.e. extmod/modnetwork.c and extmod/modsocket.c) used by the other ports.

**Reply by dpgeorge** [reply]
It might be easiest to just use the pure-Python Wiznet driver at this stage.
